source/advanced-topics.adoc

```diff
@@ -408,6 +408,145 @@ an `(opts)` field. The `opts` (the string between parentheses) is passed
 exactly as is to getopt(3) for argc/argv processing at the beginning of
 a macro invocation.
 
+[[setup]]
+=== %setup
+
+With a bunch of source code tarballs used to build the package, the mighty
+``%setup`` macro and all its advanced features might come handy. But lets begin
+with its standard behavior. You can study it from the ``rpmbuild`` ouput. At the
+beginning of each phase, it outputs ``Executing(%something)``, e.g.:
+
+[source,bash]
+----
+Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.DhddsG
+----
+
+followed by output, similar to the output of shell when ``set -x`` is enabled,
+because it is actually shell output with ``set -x`` enabled. If that is not
+enough, it is possible to peek at content of ``/var/tmp/rpm-tmp.DhddsG``. For
+that, you have to run ``rpmbuild`` with ``--debug`` option, otherwise the file
+is deleted once the build succeeds. You can see there setup of environment
+variables followed by something like:
+
+[source,bash]
+----
+cd '/builddir/build/BUILD'
+rm -rf 'cello-1.0'
+/usr/bin/gzip -dc '/builddir/build/SOURCES/cello-1.0.tar.gz' | /usr/bin/tar -xof -
+STATUS=$?
+if [ $STATUS -ne 0 ]; then
+  exit $STATUS
+fi
+cd 'cello-1.0'
+/usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+----
+
+As you can see, the ``%setup`` ensured we are working in the right directory,
+removed residues of previous builds, unpacked the source tarball and setup some
+default privileges. If we want to adjust the behavior of the ``%setup`` macro,
+we can achieve the by various options.
+
+[[setup-q]]
+==== ``%setup -q``
+
+Option ``-q`` limits verbosity of ``%setup`` macro, which mainly means that only
+``tar -xof`` is executed instead of ``tar -xvvof``. This option is listed here
+as first, since it has to be used first.
+
+[[setup-n]]
+==== ``%setup -n``
+
+Sometimes, the folder from expanded tarball has different name then the expected
+``%{name}-%{version}``. This could confuse the ``%setup`` macro, hence the
+proper name of the directory has to be specified by ``-n directory_name``
+option.
+
+For example, if the package name was ``cello``, but the source code was in
+``hello-1.0.tgz`` archive and contained ``hello/`` directory, the SPEC file
+content would need to be:
+
+[source,specfile]
+----
+Name: cello
+Source0: https://example.com/%{name}/release/hello-%{version}.tar.gz
+…
+%prep
+%setup -n hello
+----
+
+[[setup-c]]
+==== ``%setup -c``
+
+If the sourcetarball doesn't cotain any subdirectories and when unpacked, the
+files from the archive pollute the current directory, we can fix
+that by ``-c`` option. This creates the directory and step into it prior
```

Comment:
The plural verb "step" does not appear to agree with the singular subject "This". Consider changing the verb form for subject-verb agreement. (steps)